src/nvim/msgpack_rpc/channel.c

```diff
@@ -430,6 +432,13 @@ static void handle_request(Channel *channel, msgpack_object *request)
     handler.async = true;
   }
 
+  if (!handler.async) {
+    char buf[256] = { 0 };
+    memcpy(buf, method->via.bin.ptr, MIN(255, method->via.bin.size));
+    bool is_get_mode = strcmp("nvim_get_mode", buf) == 0;
+    handler.async = is_get_mode && input_blocking();
```

Comment:
special case `nvim_get_mode`: If we know RPC will be blocked, it can (and must) be asynchronous.